cpp/include/raft/matrix/detail/gather.cuh

```diff
@@ -343,6 +349,98 @@ void gather_if(const InputIteratorT in,
   typedef typename std::iterator_traits<MapIteratorT>::value_type MapValueT;
   gatherImpl(in, D, N, map, stencil, map_length, out, pred_op, transform_op, stream);
 }
+
+template <typename MatrixT, typename MapT, typename MapTransformOp, typename IndexT>
+void gatherInplaceImpl(raft::resources const& handle,
+                       raft::device_matrix_view<MatrixT, IndexT, raft::layout_c_contiguous> inout,
+                       raft::device_vector_view<const MapT, IndexT, raft::layout_c_contiguous> map,
+                       MapTransformOp transform_op,
+                       IndexT batch_size)
+{
+  // return type of MapTransformOp, must be convertible to IndexT
+  typedef typename std::result_of<decltype(transform_op)(MapT)>::type MapTransformOpReturnT;
+  RAFT_EXPECTS((std::is_convertible<MapTransformOpReturnT, IndexT>::value),
+               "MapTransformOp's result type must be convertible to signed integer");
+
+  IndexT m          = inout.extent(0);
+  IndexT n          = inout.extent(1);
+  IndexT map_length = map.extent(0);
+
+  // skip in case of 0 length input
+  if (map_length <= 0 || m <= 0 || n <= 0 || batch_size < 0) return;
+
+  RAFT_EXPECTS(map_length <= m, "Length of map should be <= number of rows for inplace gather");
```

Comment:
For gather, we can allow a map length shorter than `m`, because that won't cause race conditions. But for scatter, it should be strictly equal to `m`. The reason why I have enforced it to be strictly equal to m is again because the map should be a permutation of the range `[0, m)`. However, this condition can be relaxed to: allowing a map whose length is shorter than m and its indices are unique in the range `[0, m)`.